### ISSUE TITLE
DRG: New M365DSCDRGUtil.psm1 for reusable DRG functions

### ISSUE THREAD

**Issue by andikrueger** [microsoft/Microsoft365DSC#3309]
While browsing the code for some additional information for an issue, I came across the following functions:

https://github.com/microsoft/Microsoft365DSC/blob/Dev/Modules/Microsoft365DSC/DSCResources/MSFT_AADAdministrativeUnit/MSFT_AADAdministrativeUnit.psm1#L1057-L1743

This looks like to be part of the DRG Template. These functions should be removed from all resources and be bundled in a separate module under Modules: M365DSCDRGUtil.psm1

@William-Francillette and @NikCharlebois any thoughts on this?

**Comment by NikCharlebois** [comment]
+1

Also, all the ConvertTo- (String, complexobject, etc.) function shouldn't be found a separate functions in each resource, but instead should be defined only once in a central module.

**Comment by William-Francillette** [comment]
Sounds like a plan - I'll post that PR along with #3298 changes

**Comment by NikCharlebois** [comment]
@William-Francillette any updates on this. Where are we at with centralizing the function and removing it from all resources? Thanks

**Comment by William-Francillette** [comment]
I was delayed with the setting catalog resource generator which is still not ready.
I'll work on the cleanup separately in the week end and next week and make sure to release the pr during the week

**Comment by NikCharlebois** [comment]
Thanks for the update, appreciate your help!

**Comment by William-Francillette** [comment]
Quick update on this PR:
I finished updating the resources - I will now start the testing phase - Only x72 resource to check 😱

Will post that PR asap

**Comment by William-Francillette** [comment]
update:
Phase 1 - Unit test via Pester is completed 
There was quite more work than expected and fixed some of the resources and unit tests
@NikCharlebois @ykuijs @andikrueger I had to recreate the stubs for _MgDirectoryObject_ used by AADEntitlementManagementConnectedOrganization - Can we include Microsoft.Graph.DirectoryObjects from the dependencies?

Next phase of the test will be the resource export - I'll keep you posted

**Comment by andikrueger** [comment]
There is a dedicated module within M365DSC to create stubs: https://github.com/microsoft/Microsoft365DSC/blob/Dev/Modules/Microsoft365DSC/Modules/M365DSCStubsUtility.psm1 Looks like, this file misses some of the dependencies...

**Comment by William-Francillette** [comment]
Thanks @andikrueger - got that one, I will update with that PR

**Comment by William-Francillette** [comment]
Phase 2A - Export
I was able to export all DRG generated AAD resources successfully
It all looked ok except for:

1. A small indent issue on _AADAuthenticationMethodPolicyAuthenticator_: I will need to look a _Get-M365DSCDRGComplexTypeToString_ but this doesn't have an impact on the resource - it's only visual
2. Trailing characters: I could see some trailing characters (comma and semicolon) from CIM Instances array left by _Convert-DSCStringParamToVariable_: this should be fixed by https://github.com/microsoft/ReverseDSC/pull/33

**Comment by ricmestre** [comment]
@William-Francillette Hi, did you really encounter resources with semicolons? I really thought this wasn't true anymore so my latest iteration on that PR doesn't cover that, only single colons in an whole line. Could you please share one of those examples so I can update the PR?

**Comment by William-Francillette** [comment]
this on is for the comma
![image](https://github.com/microsoft/Microsoft365DSC/assets/79221284/fffbdf82-e69a-4337-8281-e6bb40885aee)
semicolon
![image](https://github.com/microsoft/Microsoft365DSC/assets/79221284/03b91827-2209-4657-87dc-9ad04aafafb6)
Would be great if you could include all the trailing characters as I am removing the below from all resources
![image](https://github.com/microsoft/Microsoft365DSC/assets/79221284/8ecbae00-2105-4512-a828-1f056cbb875e)

**Comment by ricmestre** [comment]
just updated the PR with this chunk:

```powershell
  if ($IsCIMArray)
  {
      $DSCBlock = $DSCBlock.Replace("},`r`n", "`}`r`n")
      #$DSCBlock = $DSCBlock.Replace("`r`n,`r`n", "`r`n") # see below
      #$DSCBlock = $DSCBlock.Replace("`r`n;`r`n", "`r`n") # see below
      $DSCBlock = $DSCBlock -replace "`r`n\s*[,;]`r`n", "`r`n" # replace "<crlf>[<whitespace>][,;]<crlf>" with "<crlf>"
  }
```

**Comment by William-Francillette** [comment]
Quick update: 

I have all tested all modified AAD resources for compilation successfully

I will be posting a PR in ReverseDSC for trailing character soon 

I have started testing update for those resources but haven't completed yet as I found some issues

Hope posting this PR this week

**Comment by William-Francillette** [comment]
AAD resources test completed 🥳

**Comment by William-Francillette** [comment]
ok just finished to merge with latest V2.0 update - I will need to retest everything as I had to resolve many conflict manually 

I hope pushing this PR in the week end ( I was nearly done lol 🥲)

**Comment by ricmestre** [comment]
@William-Francillette Really looking forward for this, great work!

I know it's not related but didn't want to create a new issue just for it, would it be possible to update the DRG to create new types of instances for Assignments as @NikCharlebois requested, e.g. MSFT_IntuneWindowsInformationProtectionPolicyWindows10MdmEnrolledAssignments instead of being hardcoded to MSFT_DeviceManagementConfigurationPolicyAssignments?

I'm asking about this because it'd really ease the work on my side because on a next step I'd like to have that object extended to accept also DisplayName and not only GroupID, right now I have my code adapted to connect to Graph in order to translate GroupId to DisplayName and vice-versa.
